### PR TITLE
Change function calls to use vectorcall

### PR DESCRIPTION
**Pull request by b-pass**
## Description

Change both the Python->C++ and the C++->Python function calls to use the faster "vector call" calling protocol.

This makes function calls faster, at the minor expense of having to handle the arguments slightly differently.

## Suggested changelog entry:

* Improved performance of function calls between Python and C++ by switching to the "vectorcall" calling protocol.

### CODE REVIEW

**Comment by b-pass** [comment]
Small benchmark results for short function calls, this PR provides modest improvement over current master... perf

**Comment by rwgk** [comment]
@b-pass I merged master, which had no conflicts.

My idea was that maybe Cursor can figure out the root cause for these two CI failures:
* https://github.com/pybind/pybind11/actions/runs/20530886225/job/58982141128?pr=5948
* https://github.com/pybind/pybind11/actions/runs/20530886225/job/58982141118?pr=5948

```
2025-12-26T23:29:10.4917170Z  E       ModuleNotFoundError: No module named 'mod_per_interpreter_gil'
...
2025-12-26T23:29:12.4779770Z  E       ModuleNotFoundError: No module named 'mod_shared_interpreter_gil'
```

I'll wait for this CI run to finish, to see if these errors appear again.

When testing locally with Python 3.14.2, default and freethreaded, I saw this error (with both builds):

```
>       m.array_reshape2(b)
E       ValueError: array_reshape2: input array total size is not a squared integer
```
**EDIT: Please see https://github.com/pybind/pybind11/pull/5948#issuecomment-3693696047 for the resolution.**

I'm attaching the full build & test logs, JIC.

I switched back to master, everything else being equal, this error didn't appear. So it's definitely somehow connected to this PR.

[pybind11_gcc_v3.14_df793163d58_default_tests_log_2025-12-26+193154.txt](https://github.com/user-attachments/files/24352325/pybind11_gcc_v3.14_df793163d58_default_tests_log_2025-12-26%2B193154.txt)

[pybind11_gcc_v3.14_df793163d58_freethreaded_log_2025-12-26+193608.txt](https://github.com/user-attachments/files/24352324/pybind11_gcc_v3.14_df793163d58_freethreaded_log_2025-12-26%2B193608.txt)

**Comment by rwgk** [comment]
Quick update: I think Cursor found the root cause for the CIBW / iOS failures, and a fix.

I'm still looking into why we're not seeing the `m.array_reshape2(b)` errors here.

**Comment by rwgk** [comment]
I pushed commit ca72d6c1a23d1fd069c5374cf5944f0932f26bc9 here because we only have very limited GHA resources. Depending on how it goes, it might be best to merge that commit in a separate PR. Let's see.

Ideally commit 6ed6d5a83ddb0bead44016d2f306ae4ab8ec8990 should also be in a separate PR, but same issue with the GHA resources.

**Comment by rwgk** [comment]
## NumPy 2.3.5 vs 2.4.0 Observations

### The Issue

When testing this PR locally with **Python 3.14 + NumPy 2.3.5**, I observed a test failure:

```
>       m.array_reshape2(b)
E       ValueError: array_reshape2: input array total size is not a squared integer

test_numpy_array.py::test_array_resize FAILED
```

This occurred when calling `array_reshape2` on a transposed view of a resized 64-element array (which should have worked since 64 = 8²).

### What We Know For Sure

1. **The error appeared with NumPy 2.3.5** (released 2025-11-16) on Python 3.14
2. **The error is gone with NumPy 2.4.0** (released 2025-12-20) on Python 3.14
3. **This is NOT a bug in the vectorcall PR** — the same test passes on master with NumPy 2.3.5, but that's because CI was skipping numpy tests for Python 3.14 entirely (no numpy entry in `tests/requirements.txt` for 3.14)
4. **NumPy 2.4.0 wheels for Python 3.14** are now available on PyPI (both default `cp314` and free-threaded `cp314t`)

### What's Undetermined

The exact bug fix in NumPy between 2.3.5 and 2.4.0 that resolved this issue. I searched the [NumPy 2.4.0 release notes](https://numpy.org/doc/stable/release/2.4.0-notes.html) but couldn't find a specific fix that clearly matches our observation. The release notes mention "many expired deprecations and bug fixes" without listing all of them individually.

### Resolution

Added `numpy==2.4.0` to `tests/requirements.txt` for Python 3.14+ (commit ca72d6c1). This:
- Enables numpy tests on Python 3.14 in CI (previously skipped)
- Avoids the NumPy 2.3.5 bug

**Comment by rwgk** [comment]
# Cursor Code Review (with some manual edits)

## Question for the Author

Is there a specific reason `unpacking_vectorcall_collector` requires Python 3.12+ instead of 3.8+?

___


## Overview

This PR changes pybind11's function calling convention from `METH_VARARGS | METH_KEYWORDS` to `METH_FASTCALL | METH_KEYWORDS` (vectorcall protocol). This is a performance optimization that avoids tuple allocation for positional arguments when calling C++ functions from Python.

The PR also adds a new `unpacking_vectorcall_collector` for the reverse direction (C++ calling Python) on Python 3.12+.

## Summary of Changes

### 1. `include/pybind11/pybind11.h` (183 lines changed)

**Key Changes:**
- Changed `ml_flags` from `METH_VARARGS | METH_KEYWORDS` to `METH_FASTCALL | METH_KEYWORDS`
- Changed dispatcher signature from `(PyObject *self, PyObject *args_in, PyObject *kwargs_in)` to `(PyObject *self, PyObject *const *args_in_arr, size_t nargsf, PyObject *kwnames_in)`
- Replaced `PyTuple_GET_ITEM(args_in, i)` with `args_in_arr[i]`
- Replaced `PyTuple_GET_SIZE(args_in)` with `PyVectorcall_NARGS(nargsf)`
- Changed kwargs handling from dict-based to tuple-based (kwnames)
- Added `keyword_index()` helper function for efficient keyword lookup
- Added `used_kwargs` vector to track which kwargs have been consumed

**Observations:**
- ✅ The `keyword_index()` function uses a two-pass approach: first checking pointer equality (for interned strings), then falling back to lexicographic comparison. This is a smart optimization.
- ✅ The code properly handles `PY_VECTORCALL_ARGUMENTS_OFFSET` in the outgoing direction.
- ⚠️ The `used_kwargs` vector is sorted before iterating to build the kwargs dict. This is O(n log n) but n is typically very small.

### 2. `include/pybind11/cast.h` (194 lines added)

**Key Changes:**
- Added `unpacking_vectorcall_collector<policy>` class for Python 3.12+ (`PY_VERSION_HEX >= 0x030C0000`)
- This class collects arguments for C++ → Python calls using the vectorcall protocol
- Uses `PyObject_Vectorcall()` instead of `PyObject_Call()`

**Observations:**
- ✅ Properly uses `PY_VECTORCALL_ARGUMENTS_OFFSET` to allow the callee to use the space before the first argument
- ✅ Keeps temporary objects alive in `m_temp` vector to prevent premature destruction
- ✅ Builds keyword names as a list first, then converts to tuple at the end (avoids resizing tuples)
- ⚠️ The code uses `PyList_Append` with `name.release().ptr()` - this transfers ownership but could leak if the append fails. However, `throw error_already_set()` is called immediately after, so this should be fine.

### 3. `include/pybind11/detail/argument_vector.h` (123 lines changed)

**Key Changes:**
- Renamed/generalized `argument_vector` to `small_vector<T, InlineSize>`
- Removed `static_assert` for trivially move constructible/destructible (now supports non-trivial types)
- Added proper destructor calls for inline storage
- Changed from `memcpy` to proper move construction
- Added `data()` method
- Added `sort()` method
- Made `argument_vector` and `args_convert_vector` type aliases to `small_vector`

**Observations:**
- ✅ The generalization allows `small_vector` to be used for other types (e.g., `small_vector<size_t, ...>` for `used_kwargs`)
- ✅ Proper handling of non-trivial types with explicit destructor calls
- ✅ Added `static_assert` for exception safety in `move_to_heap_vector_with_reserved_size`

### 4. `tests/test_kwargs_and_defaults.py` (3 lines changed)

**Key Changes:**
- Updated `test_args_refcount` to expect 3 extra references when creating a new tuple internally

**Observations:**
- ✅ This is expected behavior change: with vectorcall, the code creates a new tuple for `*args` which holds references to the arguments

## Question regarding Python Version Compatibility
- The dispatcher changes apply to all Python versions
- The `unpacking_vectorcall_collector` is only for Python 3.12+ (vectorcall was added in 3.8, but the PR uses 3.12+ for the C++ → Python direction)

**Question:** Why is `unpacking_vectorcall_collector` limited to Python 3.12+? Vectorcall has been available since Python 3.8. Perhaps there's a specific API used that requires 3.12+?

## Code Quality

### Strengths
- ✅ Well-structured changes with clear separation of concerns
- ✅ Good use of small vector optimization to avoid heap allocations
- ✅ Proper handling of edge cases (empty kwargs, etc.)
- ✅ The `keyword_index()` function is well-optimized for the common case

### Minor Suggestion

- **Comment clarity**: The comment "// -1 to account for PY_VECTORCALL_ARGUMENTS_OFFSET" could be expanded to explain that the first slot is reserved for the callee to potentially use.

## Compatibility

- **ABI**: This change affects the internal calling convention but should be transparent to users
- **API**: No public API changes
- **Behavior**: Minor reference counting changes (documented in test change)

## Testing

The existing test suite appears to cover the changes well:
- ✅ All Python 3.14 tests pass (with numpy 2.4.0)
- ✅ CI is passing across multiple platforms and Python versions
- ✅ The kwargs/args tests explicitly verify the new behavior

## Verdict

**Recommendation: Approve with minor suggestions**

This is a well-implemented performance optimization that:
1. Uses the modern vectorcall protocol for better performance
2. Properly handles all edge cases
3. Maintains backward compatibility
4. Has good test coverage

The changes are consistent with Python's own direction (vectorcall is the preferred calling convention) and should provide performance benefits, especially for frequently-called functions with few arguments.

File: include/pybind11/pybind11.h [hunk 1]
```diff
@@ -1007,46 +1007,42 @@ class cpp_function : public function {
...
-                            }
-
-                            call.args.push_back(value);
-                            call.args_convert.push_back(arg_rec.convert);
-                        } else {
-                            break;
+                        // If we're at the py::args index then first insert a stub for it to be
+                        // replaced later
+                        if (func.has_args && call.args.size() == func.nargs_pos) {
+                            call.args.push_back(none());
```
Comment (Skylion007) [inline include/pybind11/pybind11.h]: Is there no ctor to construct none from a tag we could emplace I guess?

File: include/pybind11/detail/argument_vector.h [hunk 2]
```diff
@@ -146,53 +142,73 @@ struct argument_vector {
...
             }
         } else {
             reserve_slow_path(sz);
         }
     }
 
+    void sort() {
+        if (is_inline()) {
+            if (m_repr.iarray.size < m_repr.iarray.arr.size()) {
+                std::sort(m_repr.iarray.arr.begin(),
```
Comment (Skylion007) [inline include/pybind11/detail/argument_vector.h]: Should this be sort or stable_sort?

File: include/pybind11/detail/argument_vector.h [hunk 3]
```diff
@@ -146,53 +142,73 @@ struct argument_vector {
...
         } else {
-            push_back_slow_path(x);
+            push_back_slow_path(std::move(x));
         }
     }
 
     template <typename Arg>
     void emplace_back(Arg &&x) {
-        push_back(handle(x));
+        push_back(T(x));
```
Comment (Skylion007) [inline include/pybind11/detail/argument_vector.h]:
```suggestion
        push_back(T(std::forward<Arg>(x)));
```
Handle did not have a move from ctor so this wasn't necessary.

Isn't this actually easier to do the other wya around and start with implementing emplace_bacj and have push_back explicitly call it? This ruins a lot of the advantages of emplace_back like being able to reduce the number of moves here.

File: include/pybind11/detail/argument_vector.h [hunk 4]
```diff
@@ -201,32 +217,35 @@ struct argument_vector {
...
+        static_assert(std::is_nothrow_move_assignable<T>::value,
+                      "this conversion is not exception safe");
+        static_assert(std::is_nothrow_move_constructible<heap_vector>::value,
+                      "this conversion is not exception safe");
+        std::move(ha.arr.begin(), ha.arr.begin() + ha.size, std::back_inserter(hv.vec));
         new (&m_repr.hvector) heap_vector(std::move(hv));
     }
 
-    PYBIND11_NOINLINE void push_back_slow_path(handle x) { m_repr.hvector.vec.push_back(x); }
+    PYBIND11_NOINLINE void push_back_slow_path(T x) { m_repr.hvector.vec.push_back(std::move(x)); }
```
Comment (Skylion007) [inline include/pybind11/detail/argument_vector.h]: Why not just have this fallback to emplace_back?

File: include/pybind11/cast.h [hunk 5]
```diff
@@ -2368,6 +2369,199 @@ unpacking_collector<policy> collect_arguments(Args &&...args) {
...
+    small_vector<object, arg_vector_small_size> m_temp;
+};
+
+/// Collect all arguments, including keywords and unpacking for vectorcall
+template <return_value_policy policy, typename... Args>
+unpacking_vectorcall_collector<policy> collect_arguments(Args &&...args) {
+    // Following argument order rules for generalized unpacking according to PEP 448
+    static_assert(constexpr_last<is_positional, Args...>()
+                          < constexpr_first<is_keyword_or_ds, Args...>()
+                      && constexpr_last<is_s_unpacking, Args...>()
```
Comment (Skylion007) [inline include/pybind11/cast.h]: Nit: may want to break up into two static asserts so it's easy to know which condition failed.

File: include/pybind11/detail/argument_vector.h (hunk 3 above)
Comment (b-pass) [reply include/pybind11/detail/argument_vector.h]: Sure. Done.

File: include/pybind11/cast.h (hunk 5 above)
Comment (b-pass) [reply include/pybind11/cast.h]: Done

File: include/pybind11/detail/argument_vector.h (hunk 4 above)
Comment (b-pass) [reply include/pybind11/detail/argument_vector.h]: Done.

File: include/pybind11/detail/argument_vector.h (hunk 2 above)
Comment (b-pass) [reply include/pybind11/detail/argument_vector.h]: Removed.

**Comment by rwgk** [comment]
## Review of Changes Since Commit 849df0bc

### 1. 849df0bc - argument_vector cleanups

Addressed @Skylion007's review comments:
- `[T002]`: Changed `emplace_back` to use `std::forward<Arg>(x)` for proper forwarding
- `[T003]`: Made `push_back` delegate to `emplace_back`
- `[T001]`: Removed the `sort()` method entirely (no longer needed, see below)

### 2. b517942b - Unified to vectorcall for all Python versions

This addressed the review question about why `unpacking_vectorcall_collector` was limited to Python 3.12+. The old `simple_collector` and `unpacking_collector` were removed entirely, and everything now uses vectorcall since it's been available since Python 3.8.

**Net result: -221 lines, +55 lines** — a major simplification!

### 3. b4ca3664 - `used_kwargs` as bool vector

Changed from `small_vector<size_t>` (indices of used kwargs, required sorting) to `small_vector<bool>` (flag per kwarg position). This:
- Eliminates the need for sorting
- Uses bit-packing in the specialized `small_vector<bool>` template
- More memory efficient and simpler logic

### 4. c2e65f13 and 1526a85d - Signedness fixes

Minor fixes for clang warnings about signed/unsigned conversions.

---

All changes look good. The code is cleaner and more efficient. The review comments from @Skylion007 were all addressed.

**Comment by Skylion007** [comment]
The push_back impl isnt quite right right. The pushback has an rvalue overload too. So it needs to either have both a const ref / rvalue overload (with corresponding std::move) or take by value and always move.

**Review by Skylion007 — COMMENTED** [review]
Needs a static assert the heapvector? And maybe iarray? is noexcept move constructable technically, doesn't it? Otherwise, you need some weird edge case handling for OOM like you do with std::vector

**Comment by Skylion007** [comment]
Random design question, why does the collector store pyobject pointers directly vs py::handle?

**Comment by b-pass** [comment]
> The push_back impl isnt quite right right. 

Fixed.

> Needs a static assert the heapvector? And maybe iarray?

Not sure what you mean... elaborate?

> Random design question, why does the collector store pyobject pointers directly vs py::handle?

It was to avoid having to reinterpret_cast the C++ type to the C type in the Vectorcall call....  But I've consolidated both the `PyObject*` vector and the temp `object` vector into one and casted it in this latest version.

**Comment by Skylion007** [comment]
> > The push_back impl isnt quite right right.
> 
> Fixed.
> 
> > Needs a static assert the heapvector? And maybe iarray?
> 
> Not sure what you mean... elaborate?
> 
> > Random design question, why does the collector store pyobject pointers directly vs py::handle?
> 
> It was to avoid having to reinterpret_cast the C++ type to the C type in the Vectorcall call.... But I've consolidated both the `PyObject*` vector and the temp `object` vector into one and casted it in this latest version.

Avoiding the reintrepret_cast is a valid reason, just curious.

File: include/pybind11/cast.h [hunk 6]
```diff
@@ -2292,22 +2328,31 @@ class unpacking_collector {
             throw cast_error_unable_to_convert_call_arg(a.name, a.type);
 #endif
         }
-        m_kwargs[a.name] = std::move(a.value);
+        if (PyList_Append(names_list.ptr(), name.release().ptr()) < 0) {
+            throw error_already_set();
+        }
+        m_args.emplace_back(a.value);
```
Comment (Skylion007) [inline include/pybind11/cast.h]:
```suggestion
        m_args.emplace_back(std::move(a.value));
```

File: include/pybind11/cast.h [hunk 7]
```diff
@@ -2191,94 +2190,131 @@ class argument_loader {
...
-
-    tuple args() && { return std::move(m_args); }
-    dict kwargs() && { return std::move(m_kwargs); }
-
     /// Call a Python function and pass the collected arguments
     object call(PyObject *ptr) const {
-        PyObject *result = PyObject_Call(ptr, m_args.ptr(), m_kwargs.ptr());
+        size_t nargs = m_args.size() - 1; // -1 for PY_VECTORCALL_ARGUMENTS_OFFSET (see ctor)
+        if (m_names) {
+            nargs -= static_cast<size_t>(PyTuple_GET_SIZE(m_names.ptr()));
```
Comment (Skylion007) [inline include/pybind11/cast.h]: You are assuming m_names here is a tuple when it's typed as an object.

File: include/pybind11/cast.h [hunk 8]
```diff
@@ -2191,94 +2190,131 @@ class argument_loader {
...
+        for (size_t i = 0; i < nargs; ++i) {
+            val[i] = m_args[i + 1]; // +1 for PY_VECTORCALL_ARGUMENTS_OFFSET (see ctor)
+        }
+        return val;
+    }
+
+    dict kwargs() const {
+        dict val;
+        if (m_names) {
+            auto namestup = reinterpret_borrow<tuple>(m_names);
```
Comment (Skylion007) [inline include/pybind11/cast.h]: You also assume it's a tuple here. When is m_names not a tuple? Remove this ackward explicit borrow too

File: include/pybind11/cast.h [hunk 9]
```diff
@@ -2333,39 +2378,20 @@ class unpacking_collector {
     }
 
 private:
-    tuple m_args;
-    dict m_kwargs;
+    small_vector<object, arg_vector_small_size> m_args;
+    object m_names; // null or a tuple of names
```
Comment (Skylion007) [inline include/pybind11/cast.h]:
```suggestion
    tuple m_names; // null or a tuple of names
```
Hey to break it to you but all py::object can be a null including py::tuple (it's the default moved from state). Also reduces the need for explict reintrpret_borrow and casting everywhere.

File: include/pybind11/cast.h [hunk 10]
```diff
@@ -2191,94 +2190,131 @@ class argument_loader {
...
+        }
         for (auto a : ap) {
-            args_list.append(a);
+            m_args.emplace_back(reinterpret_borrow<object>(a)); // keep alive
         }
     }
 
-    void process(list & /*args_list*/, arg_v a) {
+    // named argument
+    void process(object &names_list, arg_v a) {
```
Comment (Skylion007) [inline include/pybind11/cast.h]: Why not just type this overload as list like it was before too?

File: include/pybind11/detail/argument_vector.h [hunk 11]
```diff
@@ -146,44 +142,49 @@ struct argument_vector {
...
+    T const &operator[](std::size_t idx) const {
         assert(idx < size());
         if (is_inline()) {
             return m_repr.iarray.arr[idx];
         }
         return m_repr.hvector.vec[idx];
     }
 
-    void push_back(handle x) {
+    void push_back(T x) { emplace_back(std::move(x)); }
```
Comment (Skylion007) [inline include/pybind11/detail/argument_vector.h]: Almost equivalent, but for the sake of max efficiency. This very marginally faster if T doesn't have a move ctor and copy ctor only. Also removes a move when copying from an lvalue. Since this is an important primitive, maybe be better to be overkill on the overloads here.
```suggestion
    void push_back(const T& x) { emplace_back(x); }
    
    void push_back(T&& x) { emplace_back(std::move(x)); }
```

File: include/pybind11/detail/argument_vector.h [hunk 12]
```diff
@@ -201,32 +202,33 @@ struct argument_vector {
         using heap_vector = typename repr_type::heap_vector;
         heap_vector hv;
         hv.vec.reserve(reserved_size);
-        std::copy(ha.arr.begin(), ha.arr.begin() + ha.size, std::back_inserter(hv.vec));
+        static_assert(std::is_nothrow_move_constructible<T>::value,
+                      "this conversion is not exception safe");
+        static_assert(std::is_nothrow_move_constructible<heap_vector>::value,
+                      "this conversion is not exception safe");
```
Comment (Skylion007) [inline include/pybind11/detail/argument_vector.h:210]: Slight nit: there is a less safe inplacement assignment path here if the objects are only std::is_nothrow_move_assignable. That's probably more complicated than it's worth.

File: include/pybind11/detail/argument_vector.h [hunk 13]
```diff
@@ -192,7 +193,7 @@ struct argument_vector {
     }
 
 private:
-    using repr_type = inline_array_or_vector<handle, N>;
+    using repr_type = inline_array_or_vector<T, InlineSize>;
     repr_type m_repr;
 
     PYBIND11_NOINLINE void move_to_heap_vector_with_reserved_size(std::size_t reserved_size) {
```
Comment (Skylion007) [inline include/pybind11/detail/argument_vector.h]: As this is written except for the assertion, this has strong noexcept guarnatees vs implace move assignment

**Comment by b-pass** [comment]
> I'm a bit confused why a bunch of list and tuple objects are being stored as generic objects? We don't do anything crazy to them and they all inherit py;:object. The only different is the implicit boolean checks maybe but those can be easily over by unwrapping the ptr any time and checking against nullptr

`tuple` and `list` don't have a constructor which allows them to be constructed as `nullptr`, where as `object`'s default constructor is a nullptr. Typing these properly would make an extra allocation. IMO we can't afford the extra allocation here (in probably the hotest code path in all of pybind11). I also don't think it would be a good idea (or at least needs a lot of debate) to make a change to pybind11 tuple/list for this case.

It's just a bit inconvenient in this class ... I can certainly add some explanation in a comment about it.

**EDIT:** Turns out there is a way: `auto null_list = reinterpret_steal<list>(handle()); `

**Comment by rwgk** [comment]
@b-pass commit d828e7dfd2b0cd676becf99e1d53853a6d0322ef is meant to be a suggestion, although already fully developed. Please let me know what you think.

**Comment by b-pass** [comment]
> Please let me know what you think.

I like it.  It's less code than I thought it would be.

File: include/pybind11/cast.h [hunk 14]
```diff
@@ -2191,94 +2190,130 @@ class argument_loader {
...
+    }
 
     /// Call a Python function and pass the collected arguments
     object call(PyObject *ptr) const {
-        PyObject *result = PyObject_CallObject(ptr, m_args.ptr());
+        size_t nargs = m_args.size() - 1; // -1 for PY_VECTORCALL_ARGUMENTS_OFFSET (see ctor)
+        if (m_names) {
+            nargs -= m_names.size();
+        }
+        PyObject *result = _PyObject_Vectorcall(
```
Comment (Skylion007) [inline include/pybind11/cast.h:2246]: Slight nit could have been a py::handle

File: include/pybind11/detail/argument_vector.h [hunk 15]
```diff
@@ -326,5 +347,71 @@ struct args_convert_vector {
...
+// Container to avoid heap allocation for N or fewer booleans.
+template <size_t N>
+using args_convert_vector = small_vector<bool, N>;
+
+/// A small_vector of PyObject* that holds references and releases them on destruction.
+/// This provides explicit ownership semantics without relying on py::object's
+/// destructor, and avoids the need for reinterpret_cast when passing to vectorcall.
+template <std::size_t InlineSize>
+class ref_small_vector {
+public:
```
Comment (Skylion007) [inline include/pybind11/detail/argument_vector.h:363]: Ugh, I don't like this, but I don't see a better way of doing due to layout compatibiltiy issues with py::object. Every other method of doing this requires a copy.

File: include/pybind11/detail/argument_vector.h [hunk 16]
```diff
@@ -326,5 +347,71 @@ struct args_convert_vector {
...
+            for (std::size_t i = 0; i < m_ptrs.size(); ++i) {
+                Py_XDECREF(m_ptrs[i]);
+            }
+            m_ptrs = std::move(other.m_ptrs);
+        }
+        return *this;
+    }
+
+    /// Add a pointer, taking ownership (no incref, will decref on destruction)
+    void push_back_steal(PyObject *p) { m_ptrs.push_back(p); }
```
Comment (Skylion007) [inline include/pybind11/detail/argument_vector.h:393]:
```suggestion
    void push_back_steal(object &&o) { m_ptrs.push_back(o.release().ptr()); }
```

File: include/pybind11/cast.h [hunk 17]
```diff
@@ -2191,94 +2190,130 @@ class argument_loader {
...
-            throw cast_error_unable_to_convert_call_arg(std::to_string(args_list.size()));
+            throw cast_error_unable_to_convert_call_arg(std::to_string(m_args.size() - 1));
 #else
-            throw cast_error_unable_to_convert_call_arg(std::to_string(args_list.size()),
+            throw cast_error_unable_to_convert_call_arg(std::to_string(m_args.size() - 1),
                                                         type_id<T>());
 #endif
         }
-        args_list.append(std::move(o));
+        m_args.push_back_steal(h.ptr()); // cast returns a new reference
```
Comment (Skylion007) [inline include/pybind11/cast.h:2291]:
```suggestion
        m_args.push_back_steal(std::move(o)); // cast returns a new reference
```
Then it can keep the good old object semantics.

File: include/pybind11/cast.h [hunk 18]
```diff
@@ -2191,94 +2190,130 @@ class argument_loader {
...
 
 private:
+    // normal argument, possibly needing conversion
     template <typename T>
-    void process(list &args_list, T &&x) {
-        auto o = reinterpret_steal<object>(
-            detail::make_caster<T>::cast(std::forward<T>(x), policy, {}));
-        if (!o) {
+    void process(list & /*names_list*/, T &&x) {
+        handle h = detail::make_caster<T>::cast(std::forward<T>(x), policy, {});
```
Comment (Skylion007) [inline include/pybind11/cast.h:2282]: Converting this to a handle here makes me a bit uncomfortable. You could keep it in an object and it just use the move from ctor, the only downside is you'll have one extra call to PYXDECREF on a nullptr.

**Review by Skylion007 — COMMENTED** [review]
Not sure if you thought of this, but if you make push_back only take in objects, you can handle the reinterpret steal and reinterpret borrow using only move and release mechanics.

File: include/pybind11/cast.h (hunk 17 above)
Comment (Skylion007) [reply include/pybind11/cast.h:2291]: These can be normal objects that you just release. The borrow semantics are a bit more ackward unless you use move by value implicit construction. This might actually be cleaner.

**Comment by Skylion007** [comment]
I like how you have it done now, I just wonder if it could be clenaed up anymore with .release().ptr() semantics.

File: include/pybind11/cast.h [hunk 19]
```diff
@@ -2292,22 +2327,27 @@ class unpacking_collector {
             throw cast_error_unable_to_convert_call_arg(a.name, a.type);
 #endif
         }
-        m_kwargs[a.name] = std::move(a.value);
+        names_list.append(std::move(name));
+        m_args.push_back_borrow(a.value.ptr());
```
Comment (Skylion007) [inline include/pybind11/cast.h:2331]: Yeah both of these could just be a push_back(&&o) with a release ptr syntax and then you don't have to duplicate the logic and while the implementation is py::objects the input is still py::object. The one downside is the extra xdecref call on the nullptr stored in the py::object, but doubt that nullcheck will be a bottleneck?
```suggestion
        m_args.push_back_steal(reinterpret_borrow<object>(a.value));
```

File: include/pybind11/cast.h (hunk 17 above)
Comment (rwgk) [reply include/pybind11/cast.h:2291]: I explored what this change would look like. Here's the diff (Cursor-generated):

```diff
diff --git a/include/pybind11/cast.h b/include/pybind11/cast.h
--- a/include/pybind11/cast.h
+++ b/include/pybind11/cast.h
@@ -2279,8 +2279,9 @@ private:
     // normal argument, possibly needing conversion
     template <typename T>
     void process(list & /*names_list*/, T &&x) {
-        handle h = detail::make_caster<T>::cast(std::forward<T>(x), policy, {});
-        if (!h) {
+        auto o = reinterpret_steal<object>(
+            detail::make_caster<T>::cast(std::forward<T>(x), policy, {}));
+        if (!o) {
 #if !defined(PYBIND11_DETAILED_ERROR_MESSAGES)
             throw cast_error_unable_to_convert_call_arg(std::to_string(m_args.size() - 1));
 #else
@@ -2288,7 +2289,7 @@ private:
                                                         type_id<T>());
 #endif
         }
-        m_args.push_back_steal(h.ptr()); // cast returns a new reference
+        m_args.push_back_steal(std::move(o));
     }

     // * unpacking
@@ -2297,7 +2298,7 @@ private:
             return;
         }
         for (auto a : ap) {
-            m_args.push_back_borrow(a.ptr());
+            m_args.push_back_borrow(a);
         }
     }

@@ -2328,7 +2329,7 @@ private:
 #endif
         }
         names_list.append(std::move(name));
-        m_args.push_back_borrow(a.value.ptr());
+        m_args.push_back_borrow(a.value);
     }

     // ** unpacking
@@ -2347,7 +2348,7 @@ private:
 #endif
             }
             names_list.append(std::move(name));
-            m_args.push_back_borrow(k.second.ptr());
+            m_args.push_back_borrow(k.second);
         }
     }

diff --git a/include/pybind11/detail/argument_vector.h b/include/pybind11/detail/argument_vector.h
--- a/include/pybind11/detail/argument_vector.h
+++ b/include/pybind11/detail/argument_vector.h
@@ -389,13 +389,13 @@ public:
         return *this;
     }

-    /// Add a pointer, taking ownership (no incref, will decref on destruction)
-    void push_back_steal(PyObject *p) { m_ptrs.push_back(p); }
+    /// Add an object, taking ownership (will decref on destruction)
+    void push_back_steal(object &&o) { m_ptrs.push_back(o.release().ptr()); }

-    /// Add a pointer, borrowing (increfs now, will decref on destruction)
-    void push_back_borrow(PyObject *p) {
-        Py_XINCREF(p);
-        m_ptrs.push_back(p);
+    /// Add a handle, borrowing (increfs now, will decref on destruction)
+    void push_back_borrow(handle h) {
+        Py_XINCREF(h.ptr());
+        m_ptrs.push_back(h.ptr());
     }

     /// Add a null pointer (for PY_VECTORCALL_ARGUMENTS_OFFSET slot)
```

The change compiles and local tests pass. However, I'd prefer to keep the current implementation for these reasons:

1. **No reduction in code size** — The change is net +1 line. The `reinterpret_steal<object>(...)` wrapper and `std::move()` add verbosity that offsets the minor savings from dropping `.ptr()` at the borrow call sites.

2. **The current code is explicit and straightforward** — `ref_small_vector` is a purpose-built container for managing `PyObject*` ownership. Having it accept raw pointers makes the ownership transfer explicit at the call site. When you see `push_back_steal(h.ptr())`, it's immediately clear you're handing off a raw pointer to a container that will manage its lifetime.

3. **The abstraction doesn't simplify anything** — We still need to deal with the underlying `PyObject*` for the vectorcall API. Wrapping the interface in `object&&`/`handle` just moves the `.ptr()` and `.release()` calls inside the method, without reducing overall complexity.

I think in this case the added abstraction layer doesn't provide tangible benefits. The current approach is clear, direct, and easy to follow.

Comment (Skylion007) [reply include/pybind11/cast.h:2291]: @rwgk The main advantage here would be removing the push_back_borrow and push_back_steal to two push_back overloads

**Review by Skylion007 — APPROVED** [review]

**Comment by rwgk** [comment]
@b-pass I'm in favor of merging based on @Skylion007 and my reviews.

We could then work on the v3.0.2 release. (I could jump-start that.)

**Comment by Skylion007** [comment]
Might be enough for a minor version bump to 3.1, no?

**Comment by rwgk** [comment]
> Might be enough for a minor version bump to 3.1, no?

Not sure, maybe let's look at the changelog PR first?

There are two PRs in a holding pattern, for those we need to bump the minor version for sure (IMO):
* #5879 
* #5887
